V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-PLATFORM/MASWE-0028.md

Reviewer: Jacobo: DONE

---

> title: Sensitive Data Exposed via the User Interface

Reviewer: ```suggestion
title: Unnecessary Exposure of Sensitive Data via the User Interface
```

Reviewer: Sometimes its needed to show sensitive data, this MASWE should focus on those times that it's not needed to display the data (or at least not completely).

> title: Unnecessary Exposure of Sensitive Data via the User Interface

Author: TODO for the whole file: remove everything from the clipboard (and autocomplete) management made by the app here and leave UI management.

Reviewer: Ok!

Reviewer: Jacobo: DONE 
I have updated it removing everything from the clipboard.

Reviewer 2: Thanks! Just accepted the suggestions

id: MASWE-0028
alias: data-leak-ui

> requirement: "The app does not unnecessarily expose sensitive data through the app user interface."

Reviewer: ```suggestion
requirement: "The app does not unnecessarily expose sensitive data through its user interface."
```

platform: [android, ios]
profiles: [L2]
threat: MAS-THREAT-0028
attacks: [MAS-ATTACK-0005, MAS-ATTACK-0041, MAS-ATTACK-0043]
mappings:
  masvs-v1: [MSTG-STORAGE-7]
  masvs-v2: [MASVS-PLATFORM-3, MASVS-STORAGE-2]
  cwe: [200, 359]
  maswe-beta: [MASWE-0053]
refs:

> - https://developer.android.com/develop/ui/views/text-and-emoji/edittext#SpecifyKeyboardType

Reviewer: ```suggestion
- https://developer.android.com/develop/ui/views/touch-and-input/keyboard-input/style#Type
```

Reviewer: Link does not work (I did not check all the metadata, just links).

- https://developer.apple.com/documentation/uikit/uitextinputtraits/1624427-issecuretextentry
status: new
---

## Overview

> This weakness occurs when sensitive data, such as passwords, PINs, card numbers, or other personal information, is unnecessarily exposed through the app's user interface.

Reviewer: ```suggestion
This weakness occurs when an app exposes more sensitive data than is necessary for the user's current task through UI components or UI-mediated platform services (like the clipboard or the autocompletion feature).
```

Reviewer: Right now we mix in this MASWE data leak in screen but also in clipboard, autocompletion... I feel good to leave it as it is, at least for now, but I prefer to clarify that the clipboard or autocompletion feature are platform services on top of the UI.

> This weakness occurs when an app exposes sensitive data beyond what is required for the user's current task, or exposes required data without protections proportionate to its sensitivity, through UI components or UI-mediated platform services, such as the clipboard or autocompletion features.

Reviewer: ```suggestion
This weakness occurs when an app exposes sensitive data beyond what is required for the user's current task, or exposes required data without protections proportionate to its sensitivity through its user interface.
```

> Typical exposure paths include displaying values in cleartext when masking would suffice, allowing sensitive values to be copied to the clipboard, letting the keyboard cache or auto-correct learn from sensitive input, and not using secure text entry for secret fields. Anyone with brief access to the device or a view of its screen can then read the information.

Reviewer: ```suggestion
Typical exposure paths include displaying secrets or complete identifiers in cleartext when masking or a partial representation would suffice, using unprotected input fields for passwords or PINs, permitting unnecessary copy to the system clipboard, and enabling keyboard or IME (Input Method Editor) features that may retain or later suggest sensitive input.
```

> 

Reviewer: For weakness, should we detail edge cases when this is not a weakness? e.g., "Displaying sensitive data is not inherently a weakness when the value is required for the user's current task and is protected in proportion to its sensitivity."

Author: Good point. I think we should define "when this is a weakness" as narrowly as possible so that it's immediately clear when "it's not." Only if there's no other option should we add specific notes about edge cases. Please check the suggestion above. Does it cover what was intended to be added?

Reviewer: It does, and I also like what you propose :)

> Typical exposure paths include displaying secrets or complete identifiers in cleartext when masking or a partial representation would suffice, using unprotected input fields for passwords or PINs, permitting unnecessary copy to the system clipboard, and enabling keyboard or IME (Input Method Editor) features that may retain or later suggest sensitive input.

Author: ```suggestion
This weakness occurs when an app exposes sensitive data beyond what is required for the user's current task, or exposes required data without protections proportionate to its sensitivity, through UI components or UI-mediated platform services, such as the clipboard or autocompletion features.

Typical exposure paths include displaying secrets or complete identifiers in cleartext when masking or a partial representation would suffice, using unprotected input fields for passwords or PINs, permitting unnecessary copying to the system clipboard, and enabling keyboard or IME (Input Method Editor) features that may retain or later suggest sensitive input.
```

Reviewer: @cpholguera Nice! Seems better. Regarding the potential overlap with MASWE-0033, we can remove everything from the clipboard (and autocomplete) management made by the app here and leave UI management, what about that?

> Typical exposure paths include displaying secrets or complete identifiers in cleartext when masking or a partial representation would suffice, using unprotected input fields for passwords or PINs, permitting unnecessary copying to the system clipboard, and enabling keyboard or IME (Input Method Editor) features that may retain or later suggest sensitive input.

Reviewer: ```suggestion
The typical exposure path involves displaying secrets or sensitive information in cleartext when masking or a partial representation would suffice, (e.g., using unprotected input fields for passwords or PINs).
```

## Modes of Introduction

> - **Non-Secure Text Entry**: Building password or PIN fields without secure text entry (e.g. `isSecureTextEntry` on iOS, `textPassword` input types on Android), so the input is shown in cleartext.

Reviewer: ```suggestion
- **Non-Secure Text Entry**: Building password or PIN fields without secure text entry behavior (e.g. [`isSecureTextEntry`](https://developer.apple.com/documentation/uikit/uitextinputtraits/issecuretextentry) on iOS, [`textPassword`](https://developer.android.com/develop/ui/views/touch-and-input/keyboard-input/style#Type) input types on Android), so the input is shown in cleartext.
```

> - **Copy Allowed on Sensitive Fields**: Allowing users to copy sensitive values, placing them on the shared clipboard.

Reviewer: ```suggestion
- **Unsafe Clipboard Exposure**: Allowing sensitive values to be copied when copying the value is unnecessary, or placing them on the system clipboard automatically when it's not needed.
```

> - **Non-Secure Text Entry**: Building password or PIN fields without secure text entry behavior (e.g. [`isSecureTextEntry`](https://developer.apple.com/documentation/uikit/uitextinputtraits/issecuretextentry) on iOS, [`textPassword`](https://developer.android.com/develop/ui/views/touch-and-input/keyboard-input/style#Type) input types on Android), so the input is shown in cleartext.

Reviewer: ```suggestion
- **Non-Secure Text Entry**: Building password or PIN fields without inserting them in platform-specific fields that mask sensitive data (e.g. [`isSecureTextEntry`](https://developer.apple.com/documentation/uikit/uitextinputtraits/issecuretextentry) on iOS, [`textPassword`](https://developer.android.com/develop/ui/views/touch-and-input/keyboard-input/style#Type) input types on Android), so the input is shown in cleartext.
```

> - **Unsafe Clipboard Exposure**: Allowing sensitive values to be copied when copying the value is unnecessary, or placing them on the system clipboard automatically when it's not needed.

Reviewer: ```suggestion
```

- **Keyboard Caching and Auto-Correct**: Using input types that let the keyboard cache or auto-correct dictionaries learn sensitive input.
- **Unmasked Sensitive Values**: Displaying full values (e.g. complete card numbers or personal identifiers) where a masked or partial representation would suffice.

## Impact

> - **Compromise of Sensitive Data**: Attackers can read personal or financial information from the screen, the clipboard, or keyboard cache files, resulting in unauthorized disclosure of user data.

Reviewer: ```suggestion
- **Compromise of Sensitive Data**: Sensitive information may be observed from the screen, transferred through the clipboard, retained by an input method, or subsequently disclosed to another user.
```

> - **Keyboard Caching and Auto-Correct**: Using input types that let the keyboard cache or auto-correct dictionaries learn sensitive input.

Reviewer: ```suggestion
```

> - **Authentication or Authorization Bypass**: Attackers can capture credentials, PINs, or one-time codes exposed through the UI, resulting in unauthorized access to the user's accounts.

Reviewer: This is a bit repeated from the previous bullet point but seems also valid for me. Once the data is compromised, if the data is auth data, this can happen.

## Mitigations

> - **Use Secure Text Entry**: Configure secret fields with the platform's secure input types so the input is masked and excluded from keyboard learning.

Reviewer: ```suggestion
- **Use Secure Text Entry for Secrets**: Configure passwords, PINs, and other secrets using the platform's secure text-entry controls, such as [`textPassword` on Android Views](https://developer.android.com/develop/ui/views/touch-and-input/keyboard-input/style), [`SecureTextField` in Jetpack Compose](https://developer.android.com/develop/ui/compose/text/user-input), or [`isSecureTextEntry` on iOS](https://developer.apple.com/documentation/uikit/uitextinputtraits/issecuretextentry). Do not rely solely on custom visual masking when a platform-provided secure text-entry control is available.
```

> - **Compromise of Sensitive Data**: Sensitive information may be observed from the screen, transferred through the clipboard, retained by an input method, or subsequently disclosed to another user.

Reviewer: ```suggestion
- **Compromise of Sensitive Data**: Sensitive information may be observed from the screen and captured by other user.
```

> - **Disable Copy for Sensitive Fields**: Prevent copying of sensitive values to the clipboard where it is not essential.

Reviewer: ```suggestion
- **Restrict Clipboard Exposure**: Prevent copying sensitive values when it is not required for the intended functionality. When copying is a legitimate feature, follow the platform's [secure clipboard handling guidance on Android](https://developer.android.com/privacy-and-security/risks/secure-clipboard-handling) and mark the content using [`ClipDescription.EXTRA_IS_SENSITIVE`](https://developer.android.com/reference/android/content/ClipDescription#EXTRA_IS_SENSITIVE) to obscure clipboard previews. On iOS, use pasteboard options such as [`localOnly`](https://developer.apple.com/documentation/uikit/uipasteboard/optionskey/localonly) and [`expirationDate`](https://developer.apple.com/documentation/uikit/uipasteboard/optionskey/expirationdate) to limit cross-device propagation and content lifetime. These mechanisms reduce exposure but do not provide access control over clipboard contents.
```

> - **Disable Keyboard Caching**: Use non-caching input types for sensitive fields so keyboards and auto-correct do not retain the values.

Reviewer: ```suggestion
- **Limit Keyboard and IME Processing**: Disable unnecessary autocorrection, spell checking, suggestions, inline predictions, and personalized learning for the data that is considered sensitive. On Android, signal that dictionary-based suggestions are unnecessary using [`TYPE_TEXT_FLAG_NO_SUGGESTIONS`](https://developer.android.com/reference/android/text/InputType#TYPE_TEXT_FLAG_NO_SUGGESTIONS) and request that the IME does not retain personalized data using [`IME_FLAG_NO_PERSONALIZED_LEARNING`](https://developer.android.com/reference/android/view/inputmethod/EditorInfo#IME_FLAG_NO_PERSONALIZED_LEARNING). Treat the no-learning flag as a defense-in-depth measure rather than a guarantee because an IME may ignore it. On iOS, configure [`autocorrectionType`](https://developer.apple.com/documentation/uikit/uitextinputtraits/autocorrectiontype), [`spellCheckingType`](https://developer.apple.com/documentation/uikit/uitextinputtraits/spellcheckingtype), and [`inlinePredictionType`](https://developer.apple.com/documentation/uikit/uitextinputtraits/inlinepredictiontype) as appropriate for the sensitivity of the input.
```

Reviewer: Maybe i'm being too much detailed here, but I wanted to provide links to official documentation as we are providing recommendations

> - **Mask Displayed Values**: Show masked or partial representations of sensitive values and reveal the full value only on explicit user request.

Reviewer: ```suggestion
- **Minimize Visual Disclosure**: Display only the portion of a sensitive value required for the user's current task. When revealing the full value is necessary, reveal it only after an explicit user action (see @MASWE-0017) and for a limited duration.
```

> - **Restrict Clipboard Exposure**: Prevent copying sensitive values when it is not required for the intended functionality. When copying is a legitimate feature, follow the platform's [secure clipboard handling guidance on Android](https://developer.android.com/privacy-and-security/risks/secure-clipboard-handling) and mark the content using [`ClipDescription.EXTRA_IS_SENSITIVE`](https://developer.android.com/reference/android/content/ClipDescription#EXTRA_IS_SENSITIVE) to obscure clipboard previews. On iOS, use pasteboard options such as [`localOnly`](https://developer.apple.com/documentation/uikit/uipasteboard/optionskey/localonly) and [`expirationDate`](https://developer.apple.com/documentation/uikit/uipasteboard/optionskey/expirationdate) to limit cross-device propagation and content lifetime. These mechanisms reduce exposure but do not provide access control over clipboard contents.

Reviewer: ```suggestion
```

> - **Limit Keyboard and IME Processing**: Disable unnecessary autocorrection, spell checking, suggestions, inline predictions, and personalized learning for the data that is considered sensitive. On Android, signal that dictionary-based suggestions are unnecessary using [`TYPE_TEXT_FLAG_NO_SUGGESTIONS`](https://developer.android.com/reference/android/text/InputType#TYPE_TEXT_FLAG_NO_SUGGESTIONS) and request that the IME does not retain personalized data using [`IME_FLAG_NO_PERSONALIZED_LEARNING`](https://developer.android.com/reference/android/view/inputmethod/EditorInfo#IME_FLAG_NO_PERSONALIZED_LEARNING). Treat the no-learning flag as a defense-in-depth measure rather than a guarantee because an IME may ignore it. On iOS, configure [`autocorrectionType`](https://developer.apple.com/documentation/uikit/uitextinputtraits/autocorrectiontype), [`spellCheckingType`](https://developer.apple.com/documentation/uikit/uitextinputtraits/spellcheckingtype), and [`inlinePredictionType`](https://developer.apple.com/documentation/uikit/uitextinputtraits/inlinepredictiontype) as appropriate for the sensitivity of the input.

Reviewer: ```suggestion
```
